packages/db/tests/models/test-granule-model.js

```diff
@@ -244,11 +475,314 @@ test('GranulePgModel.upsert() will allow a running status to replace a completed
+
+  const [granuleCumulusId] = await granulePgModel.upsertWithExecutionHistory(
+    knex,
+    granule,
+    executionCumulusId
+  );
+
+  const [secondExecutionCumulusId] = await executionPgModel.create(
+    knex,
+    fakeExecutionRecordFactory()
+  );
+
+  await granulePgModel.upsertWithExecutionHistory(
+    knex,
+    granule,
+    secondExecutionCumulusId
+  );
+
+  const granuleRecord = await granulePgModel.get(
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleRecord,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+    }
+  );
+  t.deepEqual(
+    await granuleExecutionHistoryPgModel.search(
+      knex,
+      { granule_cumulus_id: granuleCumulusId }
+    ),
+    [executionCumulusId, secondExecutionCumulusId].map((executionId) => ({
+      granule_cumulus_id: Number.parseInt(granuleCumulusId, 10),
+      execution_cumulus_id: executionId,
+    }))
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() works with transaction', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    collectionCumulusId,
+    executionCumulusId,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'running',
+  });
+
+  const [granuleCumulusId] = await knex.transaction(
+    (trx) =>
+      granulePgModel.upsertWithExecutionHistory(
+        trx,
+        granule,
+        executionCumulusId
+      )
+  );
+
+  const [granuleWithHistory] = await granulePgModel.getWithExecutionHistory(
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleWithHistory,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+      execution_cumulus_ids: [executionCumulusId],
+    }
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() does not write anything if execution history upsert fails', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    collectionCumulusId,
+    executionCumulusId,
+    granuleExecutionHistoryPgModel,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'running',
+  });
+
+  const fakeGranuleExecutionHistoryPgModel = {
+    upsert: async () => {
+      throw new Error('error');
+    },
+  };
+
+  await t.throwsAsync(
+    knex.transaction(
+      (trx) =>
+        granulePgModel.upsertWithExecutionHistory(
+          trx,
+          granule,
+          executionCumulusId,
+          fakeGranuleExecutionHistoryPgModel
+        )
+    )
+  );
+
+  t.false(
+    await granulePgModel.exists(
+      knex,
+      {
+        granule_id: granule.granule_id,
+        collection_cumulus_id: collectionCumulusId,
+      }
+    )
+  );
+  t.false(
+    await granuleExecutionHistoryPgModel.exists(
+      knex,
+      {
+        execution_cumulus_id: executionCumulusId,
+      }
+    )
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() will allow a running status to replace a completed status for different execution', async (t) => {
```

Comment:
Nit: I know we're doing it in all the test titles, so this isn't this PR specifically, but  status can be `['running', 'completed', 'failed']`, and this test is testing [completed, failed] - we should probably s/completed/non-running | finished/ or something like that.

Reply:
https://github.com/nasa/cumulus/pull/2091/commits/aad6cc800caac3b0292829e502fafeae602f8ad6